### PR TITLE
[WIP] Add Wolf-Waagen API

### PR DESCRIPTION
**Pull request by tillprochaska**

### CODE REVIEW

File: app/models/report.rb [hunk 1]
```diff
@@ -11,6 +11,8 @@ class Report < ActiveRecord::Base
   has_many :variables, dependent: :destroy
   accepts_nested_attributes_for :variables
 
+  validates :hive_id, presence: true
```
Comment (roschaefer) [inline app/models/report.rb]:
```suggestion
```

File: app/lib/wolf_waagen_api/request.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  API_HOST = 'app.wolf-waagen.de'
+  API_BASE_PATH = '/api/graph/stock/'
+
+  class Request
+    attr_accessor :hive_id, :interval, :start_date, :end_date
+
+    def initialize(hive_id:, start_date:, end_date:, interval: 'hour')
+      @hive_id = hive_id
+
+      raise ArgumentError, '`start_date` and `end_date` should be instances of `DateTime`.' unless [start_date, end_date].all? DateTime
```
Comment (roschaefer) [inline app/lib/wolf_waagen_api/request.rb:17]: I believe you have some overly strict validations here. My suggestion is to write a test first, reproduce the RuntimeError and *then* think about a good solution. E.g. validate the class *or* even better you transform the given object.

File: app/lib/wolf_waagen_api/result.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      @interval = data[:pointInterval]
+
+      raise ArgumentError, '`data.series` should be a list of series data.' unless data[:series].is_a?(Array)
+
+      @raw_series = data[:series]
+    end
+
+    def series
+      @raw_series
+        .select { |series| series[:values].is_a? Array }
```
Comment (roschaefer) [inline app/lib/wolf_waagen_api/result.rb:28]: In what occasions `@raw_series` is not an Array?

File: app/lib/wolf_waagen_api/series.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      @accuracy = data[:accuracy]
+
+      raise ArgumentError, '`data.values` should be a list of series data' unless data[:values].is_a?(Array)
+
+      @values = data[:values]
+    end
+
+    def points
+      points = @values.map.with_index do |value, index|
+        Point.new(series: self, index: index, data: value)
```
Comment (roschaefer) [inline app/lib/wolf_waagen_api/series.rb:26]: *Every* point must get a reference to the `series` :thinking:

File: spec/lib/wolf_waagen_api/series.rb [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    it 'validates id' do
+      data = @valid_data.clone
+      data[:id] = nil
+
+      expect do
+        WolfWaagenApi::Series.new(result: @valid_res, data: data)
+      end.to raise_error ArgumentError
+    end
+
+    it 'should allow unitless series' do
```
Comment (roschaefer) [inline spec/lib/wolf_waagen_api/series.rb]:
```suggestion
    it 'allows unit-less series' do
```

File: spec/lib/wolf_waagen_api/series.rb [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+      data = @valid_data.clone
+      data[:unit] = nil
+
+      expect do
+        WolfWaagenApi::Series.new(result: @valid_res, data: data)
+      end.not_to raise_error
+    end
+
+    it 'should allow series without accuracy' do
```
Comment (roschaefer) [inline spec/lib/wolf_waagen_api/series.rb]:
```suggestion
    it 'allows series without accuracy' do
```

**Review by roschaefer — CHANGES_REQUESTED** [review]
Many many restrictions and validation. Try to be more lenient next time.

File: app/lib/wolf_waagen_api/result.rb (hunk 3 above)
Comment (tillprochaska) [reply app/lib/wolf_waagen_api/result.rb:28]: If there’s not data available, the API returns `nil`.

**Pull request merged**